x/market/handler/endblock_test.go

```diff
@@ -0,0 +1,213 @@
+
+	"github.com/ovrclk/akash/testutil"
+	"github.com/ovrclk/akash/x/market/types"
+)
+
+var (
+	// Create and seed the generator.
+	// Typically a non-fixed seed should be used, such as time.Now().UnixNano().
+	// Using a fixed seed will produce the same output on every run.
+	r = rand.New(rand.NewSource(550000))
+)
+
+type winnerTest struct {
+	desc      string
+	bids      []types.Bid
+	expWinner *types.Bid
+	expErr    error
+}
+
+func (w *winnerTest) testFunc(t *testing.T) {
+	winner, err := pickBidWinner(w.bids)
+	if !errors.Is(err, w.expErr) {
+		t.Errorf("returned err: %v does not match %v", err, w.expErr)
+	}
+	if w.expWinner != nil && !winner.Equals(w.expWinner.ID()) {
+		t.Errorf("unexpected winner: %#v\n%q : %v", winner, types.BidIDString(winner.BidID), winner.Price)
+		t.Logf("winner: %+v coin: %s", winner.ID(), winner.Price.String())
+	}
+}
+
+func TestBidWinner(t *testing.T) {
+	originOID := testutil.OrderID(t)
+
+	bid0 := createBid(t, originOID, int64(3))
+	bid1 := createBid(t, originOID, int64(5))
+	bid2 := createBid(t, originOID, int64(5))
+	bid3 := createBid(t, originOID, int64(4))
+	bid4 := createBid(t, originOID, int64(5))
+	bid5 := createBid(t, originOID, int64(5))
+	bid6 := createBid(t, originOID, int64(5))
+	bid7 := createBid(t, originOID, int64(8))
+	bid8 := createBid(t, originOID, int64(6))
+	bid9 := createBid(t, originOID, int64(5))
+
+	var winnerTests = []winnerTest{
+		{
+			desc:      "no bids",
+			bids:      []types.Bid{},
+			expWinner: nil,
+			expErr:    errNoBids,
+		},
+		{
+			desc:      "single bid",
+			bids:      []types.Bid{bid9},
+			expWinner: &bid9,
+		},
+		{
+			desc:      "two bids",
+			bids:      []types.Bid{bid7, bid4},
+			expWinner: &bid4,
+		},
+		{
+			desc:      "two matching bids",
+			bids:      []types.Bid{bid9, bid7, bid4},
+			expWinner: &bid4,
+		},
+		{
+			desc:      "all the same bid values",
+			bids:      []types.Bid{bid9, bid9, bid9, bid9, bid9},
+			expWinner: &bid9,
+		},
+		{
+			desc:      "two of the same",
+			bids:      []types.Bid{bid9, bid3},
+			expWinner: &bid3,
+		},
+		{
+			desc:      "multiple of the same, but one lowest value",
+			bids:      []types.Bid{bid1, bid0, bid9, bid2, bid3, bid4},
+			expWinner: &bid0,
+		},
+		{
+			desc:      "confirm last bid in list is picked",
+			bids:      []types.Bid{bid1, bid9, bid2, bid3, bid4, bid0},
+			expWinner: &bid0,
+		},
+		{
+			desc:      "two matching low bids",
+			bids:      []types.Bid{bid5, bid6, bid7, bid8},
+			expWinner: &bid6,
+		},
+	}
+
+	for _, test := range winnerTests {
+		t.Run(test.desc, test.testFunc)
+	}
+}
+
+var errExpUnevenDistribution = errors.New("this test expects uneven winning distribution")
+
+type testDist struct {
+	desc   string
+	bidNum int
+	rounds int
+	expErr error
+}
+
+func (td *testDist) testFunc(t *testing.T) {
+	originOID := testutil.OrderID(t)
+
+	distributionSpread := make(map[int]int)
+	for i := 0; i < td.bidNum; i++ {
+		distributionSpread[i] = 0
```

Comment:
I don't think that this is necessary - `distributionSpread[hithertoUnusedSlot]` will be `int`'s "zero-value" which is `0`.